jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -266,10 +288,35 @@ private void validateMemScope(ResourceScope memScope) throws IllegalStateExcepti
 	/* Intended for memSegmtOfPtrToLongArgFilter that converts the memory segment
 	 * of the passed-in pointer argument to long.
 	 */
-	private final long memSegmtOfPtrToLongArg(MemorySegment argValue) throws IllegalStateException {
-		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue);
+	private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions options) throws IllegalStateException {
+		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
 		addMemArgScope(argValue.scope());
-		return argValue.address();
+
+		long address = argValue.address();
+		/*[IF JAVA_SPEC_VERSION >= 22]*/
+		if (!argValue.isNative() && options.allowsHeapAccess()) {
+			Thread currentThread = Thread.currentThread();
+			HeapArgInfo heapArgInfo = threadHeapArgInfoMap.get(currentThread);
+			if (heapArgInfo == null) {
+				heapArgInfo = new HeapArgInfo(new Object[argLayoutArray.length],
+						new long[argLayoutArray.length], 0);
+			}
+
+			Object[] heapBases = heapArgInfo.bases();
+			long[] heapOffsets = heapArgInfo.offsets();
+			int curHeapArgIdx = heapArgInfo.curArgIdx();
+			/* Store the heap argument's base object plus its offset to these arrays
+			 * which is passed to native to extract the heap address from it via macros
+			 * given address() for the heap argument is set to zero in OpenJDK.
+			 */
+			AbstractMemorySegmentImpl heapSegment = (AbstractMemorySegmentImpl)argValue;
+			heapBases[curHeapArgIdx] = heapSegment.unsafeGetBase();
+			heapOffsets[curHeapArgIdx] = heapSegment.unsafeGetOffset();
+			address = DOWNCALL_HEAP_ARGUMENT_ID;
+			threadHeapArgInfoMap.put(currentThread, new HeapArgInfo(heapBases, heapOffsets, curHeapArgIdx + 1));
```

Comment:
Updated for use of `ThreadLocal`, etc.
```java
			HeapArgInfo info = heapArgInfo.get();
			if (info == null) {
				info = new HeapArgInfo(argLayoutArray.length);
				heapArgInfo.set(info);
			}
			AbstractMemorySegmentImpl segment = (AbstractMemorySegmentImpl) argValue;
			info.append(segment.unsafeGetBase(), segment.unsafeGetOffset());
			address = DOWNCALL_HEAP_ARGUMENT_ID;
```

Reply:
Will update all the code around related to `ThreadLocal`.

Reply:
Updated the code with `ThreadLocal` and verified in tests.